server/controllers/stock/core.js

```diff
@@ -0,0 +1,344 @@
+/**
+ * @module stock/core
+ *
+ * @description
+ * This module is responsible for handling all function utility for stock
+ *
+ * @requires lodash
+ * @requires util
+ * @requires lib/db
+ *
+ */
+
+'use strict';
+
+const _ = require('lodash');
+const moment = require('moment');
+const util = require('../../lib/util');
+const db = require('../../lib/db');
+const identifiers = require('../../config/identifiers');
+
+const flux = {
+  FROM_PURCHASE    : 1,
+  FROM_OTHER_DEPOT : 2,
+  FROM_ADJUSTMENT  : 3,
+  FROM_PATIENT     : 4,
+  FROM_SERVICE     : 5,
+  FROM_DONATION    : 6,
+  FROM_LOSS        : 7,
+  TO_OTHER_DEPOT   : 8,
+  TO_PATIENT       : 9,
+  TO_SERVICE       : 10,
+  TO_LOSS          : 11,
+  TO_ADJUSTMENT    : 12,
+};
+
+// exports
+exports.flux = flux;
+exports.getLots = getLots;
+exports.getLotsDepot = getLotsDepot;
+exports.getLotsMovements = getLotsMovements;
+
+/**
+ * @function getLots
+ *
+ * @description returns a list of lots
+ *
+ * @param {string} sql - An optional sql script of selecting in lot
+ * @param {object} params - A request query object
+ * @param {string} finalClause - An optional final clause (GROUP BY, HAVING, ...) to add to query built
+ */
+function getLots(sqlParameter, parameters, finalClauseParameter) {
+  let finalClause = finalClauseParameter;
+  const params = parameters;
+  const sql = sqlParameter || `
+        SELECT BUID(l.uuid) AS uuid, l.label, l.initial_quantity, l.unit_cost,
+            l.expiration_date, BUID(l.inventory_uuid) AS inventory_uuid, BUID(l.purchase_uuid) AS purchase_uuid, 
+            i.delay, l.entry_date, i.code, i.text, BUID(m.depot_uuid) AS depot_uuid, d.text AS depot_text,
+            CONCAT_WS('.', '${identifiers.PURCHASE_ORDER.key}', proj.abbr, p.reference) AS purchase_reference
+        FROM lot l 
+        JOIN inventory i ON i.uuid = l.inventory_uuid 
+        JOIN purchase p ON p.uuid = l.purchase_uuid
+        JOIN project proj ON proj.id = p.project_id
+        JOIN stock_movement m ON m.lot_uuid = l.uuid AND m.flux_id = ${flux.FROM_PURCHASE} 
+        JOIN depot d ON d.uuid = m.depot_uuid 
+    `;
+
+  let queryExpiration;
+  let paramExpiration;
+  let queryEntry;
+  let paramEntry;
+  const queryArray = [];
+  const paramArray = [];
+
+  if (params.uuid) {
+    params['l.uuid'] = params.uuid;
```

Comment:
I think `FilterParser` can save you ~200 lines of code here. 

I have not looked at this in depth but this is an idea of how it could be used: 

```js
const filters = new FilterParser(params);

filters.equals('uuid', 'uuid', 'l');

// this could be filters.fullText() to allow the search to be on any part of the string
filters.equals('depot_text', 'text', 'd');
filters.equals('text', 'text', 'i');
filters.equals('label', 'label', 'l');
filters.dateFrom('expiration_date_from', 'expiration_date', 'l');
filters.dateTo('expiration_date_to', 'expiration_date', 'l');
filters.dateFrom('entry_date_from', 'entry_date', 'l');
filters.dateTo('entry_date_to', 'entry_date', 'l');
filters.dateFrom('dateFrom', 'date', 'm');
filters.dateTo('dateTo', 'date', 'm');
```